ext/drreg/drreg.c

```diff
@@ -719,24 +1137,66 @@ drreg_forward_analysis(void *drcontext, instr_t *start)
  */
 
 drreg_status_t
-drreg_init_and_fill_vector(drvector_t *vec, bool allowed)
+drreg_init_and_fill_vector_ex(drvector_t *vec, drreg_spill_class_t spill_class,
+                              bool allowed)
 {
     reg_id_t reg;
+    uint size;
     if (vec == NULL)
         return DRREG_ERROR_INVALID_PARAMETER;
-    drvector_init(vec, DR_NUM_GPR_REGS, false /*!synch*/, NULL);
-    for (reg = 0; reg < DR_NUM_GPR_REGS; reg++)
+
+    if (spill_class == DRREG_GPR_SPILL_CLASS) {
+        size = DR_NUM_GPR_REGS;
+    } else if (spill_class == DRREG_SIMD_XMM_SPILL_CLASS ||
+               spill_class == DRREG_SIMD_YMM_SPILL_CLASS ||
+               spill_class == DRREG_SIMD_ZMM_SPILL_CLASS) {
+#ifdef X86
+        size = DR_NUM_SIMD_VECTOR_REGS;
+#else
+        /* FIXME i#3844: NYI on ARM */
+        return DRREG_ERROR;
+#endif
+    } else {
+        return DRREG_ERROR;
+    }
+
+    drvector_init(vec, size, false /*!synch*/, NULL);
+    for (reg = 0; reg < size; reg++)
         drvector_set_entry(vec, reg, allowed ? (void *)(ptr_uint_t)1 : NULL);
     return DRREG_SUCCESS;
 }
 
+drreg_status_t
+drreg_init_and_fill_vector(drvector_t *vec, bool allowed)
+{
+    return drreg_init_and_fill_vector_ex(vec, DRREG_GPR_SPILL_CLASS, allowed);
+}
+
 drreg_status_t
 drreg_set_vector_entry(drvector_t *vec, reg_id_t reg, bool allowed)
 {
-    if (vec == NULL || reg < DR_REG_START_GPR || reg > DR_REG_STOP_GPR)
-        return DRREG_ERROR_INVALID_PARAMETER;
-    drvector_set_entry(vec, reg - DR_REG_START_GPR,
-                       allowed ? (void *)(ptr_uint_t)1 : NULL);
+    reg_id_t start_reg;
+
+    if (reg_is_gpr(reg)) {
+        if (vec == NULL || reg < DR_REG_START_GPR || reg > DR_REG_STOP_GPR)
+            return DRREG_ERROR_INVALID_PARAMETER;
+        start_reg = DR_REG_START_GPR;
+    }
+#ifdef X86
+    else if (reg_is_vector_simd(reg)) {
+        /* We assume the SIMD range is contiguous and no further out of range checks
+         * are performed.*/
```

Comment:
?  Why is it not checking for illegal values just like for GPR?

Reply:
I did that. Because as the comment is saying, the SIMD range is contiguous and reg_is_vector_simd() implies that reg is within that exact range.

Reply:
I see, that was easy to miss as a reader that reg_is_vector_simd is doing the check: best to include that in the comment.  Also, please fix the comment style (missing space at end; prefer `*/` on own line).

Reply:
It says "and no further out of range checks are performed". I re-phrased. Done.